lldb/source/Plugins/Process/scripted/ScriptedThread.cpp

```
@@ -204,9 +209,64 @@ bool ScriptedThread::LoadArtificialStackFrames() {
     SymbolContext sc;
     symbol_addr.CalculateSymbolContext(&sc);
 
-    StackFrameSP synth_frame_sp = std::make_shared<StackFrame>(
+    return std::make_shared<StackFrame>(
         this->shared_from_this(), idx, idx, cfa, cfa_is_valid, pc,
         StackFrame::Kind::Artificial, behaves_like_zeroth_frame, &sc);
+  };
+
+  auto create_frame_from_script_object =
+      [this, arr_sp](size_t idx) -> llvm::Expected<StackFrameSP> {
+    Status error;
+    StructuredData::ObjectSP object_sp = arr_sp->GetItemAtIndex(idx);
+    if (!object_sp || !object_sp->GetAsGeneric()) {
+      ScriptedInterface::ErrorWithMessage<bool>(
+          LLVM_PRETTY_FUNCTION,
+          llvm::Twine("Couldn't get artificial stackframe object at index (" +
+                      llvm::Twine(idx) +
+                      llvm::Twine(") from stackframe array."))
+              .str(),
+          error, LLDBLog::Thread);
+      return error.ToError();
+    }
+
+    auto frame_or_error =
+        ScriptedFrame::Create(*this, nullptr, object_sp->GetAsGeneric());
+
+    if (!frame_or_error) {
+      ScriptedInterface::ErrorWithMessage<bool>(
+          LLVM_PRETTY_FUNCTION, toString(frame_or_error.takeError()), error);
+      return error.ToError();
+    }
+
+    StackFrameSP frame_sp = frame_or_error.get();
+    lldbassert(frame_sp && "Couldn't initialize scripted frame.");
+
+    return frame_sp;
+  };
+
+  StackFrameListSP frames = GetStackFrameList();
+
+  for (size_t idx = 0; idx < arr_size; idx++) {
+    StackFrameSP synth_frame_sp = nullptr;
+
+    auto frame_from_dict_or_err = create_frame_from_dict(idx);
```

This doesn't seem right.  There are two reasons create_frame_from_dict could fail:
1) Because the frame isn't described by a dictionary but by a script object
2) Because the dict existed but was incorrect
In case 1) you do want to try the scripted object instead, but not if there was a dictionary but it failed.  Falling back to the script object and obscuring the fact that there was a bogus dictionary will be confusing.

---

I would expect that if there IS a dictionary provider that returns an actual error handling that index, we should report the error.  People would want to know that there was this dictionary provider floating around possibly interfering with the synthetic provider they have just added...
If the dictionary provider says it isn't in the business of providing that frame, that shouldn't be an error since that just means "that index is not my job".  I presume it would do that by returning an empty frame but not an error?
If the dictionary provider says "that index wasn't something I wanted to provide a frame for", you should go on and consult the synthetic one.
Also, if we are forced to handle the case where there are both synthetic and dictionary providers at the same time, we should do the synthetic one first.  After all, since that's the newer feature, it makes more sense to let it override the older variant than the other way round...

---

Right, the main thing that was bugging me was if this bit:

```
    if (!dict->GetValueForKeyAsInteger("pc", pc)) {
      ScriptedInterface::ErrorWithMessage<bool>(
          LLVM_PRETTY_FUNCTION,
          "Couldn't find value for key 'pc' in stackframe dictionary.", error,
           LLDBLog::Thread);
      return error.ToError();
    }
```

fails, you really don't want to do more work.  After all, this ScriptedThread implementation provided you what it thought was a legit dictionary entry for a frame, but it was in fact bogus.  In that case, you shouldn't try to do any more work, you should just report the error.

Internally there's just one array that either has the dictionary or the script object, so you won't lose the error.  But you also won't succeed in getting the script object out of it either because you already found a Dictionary entry there.

So the code would be cleaner if you made a distinction between "there's something other than a Dictionary in this slot" and "there's a bad Dictionary in this slot".

And even if we changed the implementation in the future so that the Dictionaries and the Script Objects didn't live in the same array, this would continue to be the right behavior.  If the ScriptedThread claims to return a Dictionary representation for that index but it isn't correct, we should report its badness rather than trying the Scripted Frame version.